src/ripple/basics/SlabAllocator.h

```diff
@@ -0,0 +1,235 @@
+    std::string const name_;
+
+    // The number of items to allocate:
+    std::size_t const count_;
+
+    // The underlying memory block:
+    std::uint8_t* p_;
+
+    // A linked list of free buffers:
+    std::uint8_t* l_ = nullptr;
+
+    // A mutex to protect the linked list:
+    std::mutex mutable m_;
+
+    // How many allocations we've serviced:
+    std::atomic<std::uint64_t> alloc_count_ = 0;
+
+    // How many of those were from our internal buffer:
+    std::atomic<std::uint64_t> alloc_fast_count_ = 0;
+
+    // How many of those went to our internal buffer:
+    std::atomic<std::uint64_t> dealloc_fast_count_ = 0;
+
+    // Scrubs the memory range [ptr, ptr+Size)
+    std::uint8_t*
+    scrub(std::uint8_t* ptr, [[maybe_unused]] std::uint8_t value)
+    {
+        assert(ptr != nullptr);
+
+#ifdef SLAB_SCRUB_MEMORY
+        std::memset(ptr, value, Size);
+#endif
+
+        return ptr;
+    }
+
+public:
+    // We may need to adjust the size of an individual item to account for
+    // alignment issues:
+    constexpr static std::size_t const size = []() constexpr {
+        std::size_t ret = sizeof(Type) + ExtraSize;
+
+        if (auto const mod = (sizeof(Type) + ExtraSize) % alignof(Type))
+            ret += (alignof(Type) - mod);
+
+        return ret;
+    }();
+
+    /** Constructs a slab allocator able to allocate up to count items
+
+        @param count the number of items the slab allocator can allocate; note
+                     that a count of 0 is valid and means that the allocator
+                     is, effectively, disabled. This can be very useful in some
+                     contexts (e.g. when mimimal memory usage is needed) and
+                     allows for graceful failure.
+     */
+    SlabAllocator(
+        std::size_t count,
+        std::string name = "Slab: " + beast::type_name<Type>())
+        : name_(name), count_(count)
+    {
+        if (count != 0)
+        {
+            p_ = reinterpret_cast<std::uint8_t*>(
+                std::aligned_alloc(alignof(Type), size * count_));
+            assert(p_ != nullptr);
+        }
+
+        // We don't treat a memory allocation
+        if (p_ != nullptr)
+        {
+            for (std::size_t i = 0; i != count_; ++i)
+                free(p_ + (i * size));
+
+            // Ignore the use of free above; we just use the API to easily
+            // "build" the free list, but it's not an actual deallocation.
+            dealloc_fast_count_ = 0;
+        }
+    }
+
+    ~SlabAllocator()
+    {
+        if (p_ != nullptr)
+            std::free(p_);
+    }
+
+    /** Returns the number of items that the allocator can accomodate. */
+    std::size_t
+    count() const
+    {
+        return count_;
+    }
+
+    /** Determines whether the given pointer belongs to this allocator */
+    bool
+    own(std::uint8_t const* p) const
+    {
+        return (p >= p_) && (p < p_ + (size * count_));
```

Comment:
Technically this is undefined behavior if `!own` because you can't compare pointers to different objects.  However from a practical standpoint, this works, and I know of no other way to do it.  I've written code like this myself and used it widely (in libc++).  I note this just so that we go into it with eyes wide open.